File: services/api/src/modules/curriculum/types/lesson-content-block.ts
export enum LessonContentBlockType {
  HEADING = 'heading',
  PARAGRAPH = 'paragraph',
  FORMULA = 'formula',
  EXAMPLE = 'example',
  IMPORTANT_NOTE = 'important_note',
  IMAGE = 'image',
  TABLE = 'table',
  CITATION = 'citation',
  LIST = 'list',
  QUOTE = 'quote',
}

export enum LessonImportantNoteSeverity {
  INFO = 'info',
  WARNING = 'warning',
  TIP = 'tip',
}

export interface LessonContentBlockBase {
  id: string;
  type: LessonContentBlockType;
  order: number;
}

export interface LessonHeadingContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.HEADING;
  text: string;
  level: 1 | 2 | 3;
}

export interface LessonParagraphContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.PARAGRAPH;
  text: string;
}

export interface LessonFormulaContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.FORMULA;
  expression: string;
  description?: string;
}

export interface LessonExampleContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.EXAMPLE;
  title?: string;
  body: string;
  solution?: string;
}

export interface LessonImportantNoteContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.IMPORTANT_NOTE;
  title?: string;
  text: string;
  severity: LessonImportantNoteSeverity;
}

export interface LessonImageContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.IMAGE;
  url: string;
  altText: string;
  caption?: string;
}

export interface LessonTableContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.TABLE;
  headers: string[];
  rows: string[][];
}

export interface LessonCitationContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.CITATION;
  bookName: string;
  chapter?: string;
  page?: string;
  excerpt?: string;
}

export interface LessonListContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.LIST;
  items: string[];
  ordered?: boolean;
}

export interface LessonQuoteContentBlock extends LessonContentBlockBase {
  type: LessonContentBlockType.QUOTE;
  text: string;
  attribution?: string;
}

export type LessonContentBlock =
  | LessonHeadingContentBlock
  | LessonParagraphContentBlock
  | LessonFormulaContentBlock
  | LessonExampleContentBlock
  | LessonImportantNoteContentBlock
  | LessonImageContentBlock
  | LessonTableContentBlock
  | LessonCitationContentBlock
  | LessonListContentBlock
  | LessonQuoteContentBlock;

export const LESSON_CONTENT_BLOCK_MAX_COUNT = 200;
export const LESSON_CONTENT_BLOCK_MAX_TABLE_ROWS = 100;
export const LESSON_CONTENT_BLOCK_MAX_TABLE_COLUMNS = 20;
export const LESSON_CONTENT_BLOCK_MAX_TEXT_LENGTH = 10000;
export const LESSON_CONTENT_BLOCK_MAX_SHORT_TEXT_LENGTH = 2000;
export const LESSON_CONTENT_BLOCK_MAX_IMAGE_URL_LENGTH = 2048;

function compareBlocks(
  a: LessonContentBlock,
  b: LessonContentBlock,
  indexA: number,
  indexB: number,
): number {
  const orderDiff = (a.order ?? 0) - (b.order ?? 0);
  if (orderDiff !== 0) {
    return orderDiff;
  }

  const idDiff = (a.id ?? '').localeCompare(b.id ?? '');
  if (idDiff !== 0) {
    return idDiff;
  }

  return indexA - indexB;
}

export function sortLessonContentBlocks(
  blocks: LessonContentBlock[] | undefined | null,
): LessonContentBlock[] {
  if (!Array.isArray(blocks) || blocks.length === 0) {
    return [];
  }

  return [...blocks]
    .map((block, index) => ({ block, index }))
    .sort((left, right) => compareBlocks(left.block, right.block, left.index, right.index))
    .map(({ block }) => block);
}

function isNonEmptyString(value: unknown, maxLength: number): value is string {
  return typeof value === 'string' && value.trim().length > 0 && value.trim().length <= maxLength;
}

function isStringArray(value: unknown, maxLength: number, maxCount: number): value is string[] {
  return (
    Array.isArray(value) &&
    value.length <= maxCount &&
    value.every((item) => isNonEmptyString(item, maxLength))
  );
}

function isStringMatrix(
  value: unknown,
  maxLength: number,
  maxRows: number,
  maxColumns: number,
): value is string[][] {
  return (
    Array.isArray(value) &&
    value.length <= maxRows &&
    value.every(
      (row) =>
        Array.isArray(row) &&
        row.length <= maxColumns &&
        row.every((cell) => isNonEmptyString(cell, maxLength)),
    )
  );
}

export function validateLessonContentBlock(value: unknown): value is LessonContentBlock {
  if (!value || typeof value !== 'object') {
    return false;
  }

  const block = value as Record<string, unknown>;
  if (
    !isNonEmptyString(block.id, 120) ||
    typeof block.order !== 'number' ||
    !Number.isFinite(block.order) ||
    block.order < 0 ||
    !isNonEmptyString(block.type, 32)
  ) {
    return false;
  }

  switch (block.type) {
    case LessonContentBlockType.HEADING:
      return (
        isNonEmptyString(block.text, LESSON_CONTENT_BLOCK_MAX_SHORT_TEXT_LENGTH) &&
        [1, 2, 3].includes(Number(block.level))
      );
    case LessonContentBlockType.PARAGRAPH:
      return isNonEmptyString(block.text, LESSON_CONTENT_BLOCK_MAX_TEXT_LENGTH);
    case LessonContentBlockType.FORMULA:
      return isNonEmptyString(block.expression, LESSON_CONTENT_BLOCK_MAX_SHORT_TEXT_LENGTH);
    case LessonContentBlockType.EXAMPLE:
      return (
        isNonEmptyString(block.body, LESSON_CONTENT_BLOCK_MAX_TEXT_LENGTH) &&
        (block.title == null ||
          isNonEmptyString(block.title, LESSON_CONTENT_BLOCK_MAX_SHORT_TEXT_LENGTH)) &&
        (block.solution == null ||
          isNonEmptyString(block.solution, LESSON_CONTENT_BLOCK_MAX_TEXT_LENGTH))
      );
    case LessonContentBlockType.IMPORTANT_NOTE:
      return (
        isNonEmptyString(block.text, LESSON_CONTENT_BLOCK_MAX_TEXT_LENGTH) &&
        Object.values(LessonImportantNoteSeverity).includes(
          block.severity as LessonImportantNoteSeverity,
        )
      );
    case LessonContentBlockType.IMAGE:
      return (
        isNonEmptyString(block.url, LESSON_CONTENT_BLOCK_MAX_IMAGE_URL_LENGTH) &&
        isNonEmptyString(block.altText, LESSON_CONTENT_BLOCK_MAX_SHORT_TEXT_LENGTH)
      );
    case LessonContentBlockType.TABLE:
      return (
        isStringArray(block.headers, 100, LESSON_CONTENT_BLOCK_MAX_TABLE_COLUMNS) &&
        isStringMatrix(
          block.rows,
          1000,
          LESSON_CONTENT_BLOCK_MAX_TABLE_ROWS,
          LESSON_CONTENT_BLOCK_MAX_TABLE_COLUMNS,
        ) &&
        block.rows.every((row) => row.length === (block.headers as string[]).length)
      );
    case LessonContentBlockType.CITATION:
      return isNonEmptyString(block.bookName, LESSON_CONTENT_BLOCK_MAX_SHORT_TEXT_LENGTH);
    case LessonContentBlockType.LIST:
      return isStringArray(block.items, LESSON_CONTENT_BLOCK_MAX_TEXT_LENGTH, 50);
    case LessonContentBlockType.QUOTE:
      return isNonEmptyString(block.text, LESSON_CONTENT_BLOCK_MAX_TEXT_LENGTH);
    default:
      return false;
  }
}

export function validateLessonContentBlocks(blocks: unknown): blocks is LessonContentBlock[] {
  return (
    Array.isArray(blocks) &&
    blocks.length <= LESSON_CONTENT_BLOCK_MAX_COUNT &&
    blocks.every((block) => validateLessonContentBlock(block))
  );
}

export function normalizeLessonContentBlocks(
  blocks: LessonContentBlock[] | undefined | null,
): LessonContentBlock[] {
  return sortLessonContentBlocks(blocks ?? []).map((block) => ({ ...block }));
}

File: apps/mobile/lib/features/curriculum/domain/entities/lesson_content_block.dart
enum LessonContentBlockType {
  heading,
  paragraph,
  formula,
  example,
  importantNote,
  image,
  table,
  citation,
  list,
  quote,
}

extension LessonContentBlockTypeApiValue on LessonContentBlockType {
  String get apiValue => switch (this) {
        LessonContentBlockType.heading => 'heading',
        LessonContentBlockType.paragraph => 'paragraph',
        LessonContentBlockType.formula => 'formula',
        LessonContentBlockType.example => 'example',
        LessonContentBlockType.importantNote => 'important_note',
        LessonContentBlockType.image => 'image',
        LessonContentBlockType.table => 'table',
        LessonContentBlockType.citation => 'citation',
        LessonContentBlockType.list => 'list',
        LessonContentBlockType.quote => 'quote',
      };
}

LessonContentBlockType lessonContentBlockTypeFromApiValue(String value) {
  return switch (value) {
    'heading' => LessonContentBlockType.heading,
    'paragraph' => LessonContentBlockType.paragraph,
    'formula' => LessonContentBlockType.formula,
    'example' => LessonContentBlockType.example,
    'important_note' => LessonContentBlockType.importantNote,
    'image' => LessonContentBlockType.image,
    'table' => LessonContentBlockType.table,
    'citation' => LessonContentBlockType.citation,
    'list' => LessonContentBlockType.list,
    'quote' => LessonContentBlockType.quote,
    _ => LessonContentBlockType.paragraph,
  };
}

enum LessonImportantNoteSeverity {
  info,
  warning,
  tip,
}

extension LessonImportantNoteSeverityApiValue on LessonImportantNoteSeverity {
  String get apiValue => switch (this) {
        LessonImportantNoteSeverity.info => 'info',
        LessonImportantNoteSeverity.warning => 'warning',
        LessonImportantNoteSeverity.tip => 'tip',
      };
}

LessonImportantNoteSeverity lessonImportantNoteSeverityFromApiValue(
  String value,
) {
  return switch (value) {
    'info' => LessonImportantNoteSeverity.info,
    'warning' => LessonImportantNoteSeverity.warning,
    'tip' => LessonImportantNoteSeverity.tip,
    _ => LessonImportantNoteSeverity.info,
  };
}

abstract class LessonContentBlock {
  final String id;
  final LessonContentBlockType type;
  final int order;

  const LessonContentBlock({
    required this.id,
    required this.type,
    required this.order,
  });

  Map<String, dynamic> toJson();
}

class LessonHeadingContentBlock extends LessonContentBlock {
  final String text;
  final int level;

  const LessonHeadingContentBlock({
    required super.id,
    required super.order,
    required this.text,
    required this.level,
  }) : super(type: LessonContentBlockType.heading);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'text': text,
        'level': level,
      };
}

class LessonParagraphContentBlock extends LessonContentBlock {
  final String text;

  const LessonParagraphContentBlock({
    required super.id,
    required super.order,
    required this.text,
  }) : super(type: LessonContentBlockType.paragraph);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'text': text,
      };
}

class LessonFormulaContentBlock extends LessonContentBlock {
  final String expression;
  final String? description;

  const LessonFormulaContentBlock({
    required super.id,
    required super.order,
    required this.expression,
    this.description,
  }) : super(type: LessonContentBlockType.formula);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'expression': expression,
        'description': description,
      };
}

class LessonExampleContentBlock extends LessonContentBlock {
  final String? title;
  final String body;
  final String? solution;

  const LessonExampleContentBlock({
    required super.id,
    required super.order,
    this.title,
    required this.body,
    this.solution,
  }) : super(type: LessonContentBlockType.example);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'title': title,
        'body': body,
        'solution': solution,
      };
}

class LessonImportantNoteContentBlock extends LessonContentBlock {
  final String? title;
  final String text;
  final LessonImportantNoteSeverity severity;

  const LessonImportantNoteContentBlock({
    required super.id,
    required super.order,
    this.title,
    required this.text,
    required this.severity,
  }) : super(type: LessonContentBlockType.importantNote);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'title': title,
        'text': text,
        'severity': severity.apiValue,
      };
}

class LessonImageContentBlock extends LessonContentBlock {
  final String url;
  final String altText;
  final String? caption;

  const LessonImageContentBlock({
    required super.id,
    required super.order,
    required this.url,
    required this.altText,
    this.caption,
  }) : super(type: LessonContentBlockType.image);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'url': url,
        'altText': altText,
        'caption': caption,
      };
}

class LessonTableContentBlock extends LessonContentBlock {
  final List<String> headers;
  final List<List<String>> rows;

  const LessonTableContentBlock({
    required super.id,
    required super.order,
    required this.headers,
    required this.rows,
  }) : super(type: LessonContentBlockType.table);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'headers': headers,
        'rows': rows,
      };
}

class LessonCitationContentBlock extends LessonContentBlock {
  final String bookName;
  final String? chapter;
  final String? page;
  final String? excerpt;

  const LessonCitationContentBlock({
    required super.id,
    required super.order,
    required this.bookName,
    this.chapter,
    this.page,
    this.excerpt,
  }) : super(type: LessonContentBlockType.citation);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'bookName': bookName,
        'chapter': chapter,
        'page': page,
        'excerpt': excerpt,
      };
}

class LessonListContentBlock extends LessonContentBlock {
  final List<String> items;
  final bool ordered;

  const LessonListContentBlock({
    required super.id,
    required super.order,
    required this.items,
    this.ordered = false,
  }) : super(type: LessonContentBlockType.list);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'items': items,
        'ordered': ordered,
      };
}

class LessonQuoteContentBlock extends LessonContentBlock {
  final String text;
  final String? attribution;

  const LessonQuoteContentBlock({
    required super.id,
    required super.order,
    required this.text,
    this.attribution,
  }) : super(type: LessonContentBlockType.quote);

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'type': type.apiValue,
        'order': order,
        'text': text,
        'attribution': attribution,
      };
}

List<LessonContentBlock> sortLessonContentBlocks(
  Iterable<LessonContentBlock> blocks,
) {
  final indexed = blocks.toList().asMap().entries.toList();
  indexed.sort((left, right) {
    final orderDiff = left.value.order.compareTo(right.value.order);
    if (orderDiff != 0) {
      return orderDiff;
    }
    final idDiff = left.value.id.compareTo(right.value.id);
    if (idDiff != 0) {
      return idDiff;
    }
    return left.key.compareTo(right.key);
  });
  return indexed.map((entry) => entry.value).toList(growable: false);
}
